CIPs/CIP-124/cip-124.md

```diff
@@ -0,0 +1,321 @@
+your_events:  {"ape", "bee", "cat", "doe", "eel", "fox", "gnu", "hog"}
+
+you: -["ape", h("bee", "cat", "doe", "eel", "fox", "hog", "gnu"), "ape"]->
+
+```
+
+In this example only `"fox"` was never sent across the wire but in large sets with only tiny fractions of diffrences most keys will never be sent across the wire.
+> ToDo: should we merge 
+> ```
+> ["ape", h("bee", "cat", "doe", "eel", "fox", "hog", "gnu"), "ape"]
+> ```
+>  to
+> ```
+> [h("ape", "bee", "cat", "doe", "eel", "fox", "hog", "gnu")]
+> ```
+> to represent the entire ring?
+
+---
+#### **Imutable object store**
+For syncing with an imutable object store add CIDs to the inodes. This would be DAG-JSON or DAG-CBOR
+```
+(EventID (sub-tree-ahash sub-tree-CID EventID)+ )+
+```
+This protocal can be run in a symalar way agenst a object store.
+however insted of the remote desiding which segments to send they pre-partion the set into a b-tree. At each laer of the b-tree you can deside if you have a match or if you need to recurse into the sub-tree.
+
+The use of synconising agenst a object store may improve the expereance of syncing a brand new node or a node that has been down for an extended pereod of time. The node could start by syncronising agenst a weekly snapshot and the sync with live nodes only after getting the bulk of the dataset from the object store.
+
+#### **DHT key**
+
+- If `separatorKey` and `separatorValue` are set, use `partition = hash(separatorKey | separatorValue)`
+- Else if `stream` is set, use `stream`
+
+### **Random Peer Synchronization Order**
+When synchronizing we want to sync with nearby nodes with higher probability since this is lower cost but also sometimes with the far away nodes so that the whole network is synchronized. By tracking the latency of other nodes with overlapping interests we can choose a peer proportional to the overlapping interests and inversely proportional to the latency.
+|| keys in overlapping interests || / latency
+
+![Random Gosip](assets/gossip.gif)
+
+The number of nodes with the event will grow exponetialy until it saturates the network.
+
+### **Associative Hash Function (SumOfShas)**
+If we want to caculate the hash of a range of 
+EventIDs without needing to visit all of the strings 
+we can use an associative hash function where the grouping
+of EventIDs into subranges is not relivent to the final hash.
+
+e.g. use `sum(sha256(EventID) for EventID in Stream Set)`
+each node in the tree can store the associative hash of each 
+sub tree along with its max and min we only need to desend into that sub-tree if the reange end is in the sub-tree. Since a range has only two ends, start and stop, we can calulate the associative hash by visiting only twice the depth of the tree number of nodes `2*log_b(n)`
+where b is the fanout and n is the number of EventIDs.
+
+A b-tree with fanout 3:
+![fanout3](assets/b_hash_tree_1.png)
+
+A b-tree with fanout 2:
+![fanout2](assets/b_hash_tree_2.png)
+
+### **B#tree (B hash trees)**
```

Comment:
Not sure I follow the context of B-trees here. I didn't see them mentioned elsewhere?